src/visitors/sympy_replace_solutions_visitor.cpp

```diff
@@ -0,0 +1,354 @@
+            return statement.substr(0, 3) != "tmp";
+        });
+    tmp_statements_ = SolutionSorter(solutions.begin(), ss_tmp_delimeter);
+    solutions_ = SolutionSorter(ss_tmp_delimeter, solutions.end());
+
+    replacements_.clear();
+    is_statement_block_root_ = true;
+}
+
+
+void SympyReplaceSolutionsVisitor::visit_statement_block(ast::StatementBlock& node) {
+    const bool is_root = is_statement_block_root_;
+    is_statement_block_root_ = false;
+
+    if (is_root) {
+        logger->debug("SympyReplaceSolutionsVisitor :: visit matching statements by value");
+        interleaves_counter_ = InterleavesCounter();
+        policy_ = ReplacePolicy::VALUE;
+        node.visit_children(*this);
+
+        if (!solutions_.is_all_untagged()) {
+            logger->debug(
+                "SympyReplaceSolutionsVisitor :: not all solutions were replaced. Visit replacing "
+                "greedly");
+            interleaves_counter_ = InterleavesCounter();
+            policy_ = ReplacePolicy::GREEDY;
+            node.visit_children(*this);
+            if (interleaves_counter_.n() > 0) {
+                logger->warn(
+                    "SympyReplaceSolutionsVisitor :: Found ambiguous system of equations "
+                    "interleaved with {} assignment statements. I do not know what equations go "
+                    "before and what "
+                    "equations go after the assignment statements. Either put all the equations "
+                    "that need to be solved "
+                    "in the form: x = f(...) and with distinct variable assignments or do not "
+                    "interleave the system with assignments.",
+                    interleaves_counter_.n());
+            }
+        }
+    } else {
+        node.visit_children(*this);
+    }
+
+
+    const auto& old_statements = node.get_statements();
+
+    ast::StatementVector new_statements;
+    new_statements.reserve(2 * old_statements.size());
+    for (const auto& old_statement: old_statements) {
+        const auto& replacement_ptr = replacements_.find(old_statement);
+        if (replacement_ptr != replacements_.end()) {
+            new_statements.insert(new_statements.end(),
+                                  replacement_ptr->second.begin(),
+                                  replacement_ptr->second.end());
+            logger->debug("SympyReplaceSolutionsVisitor :: erasing {}", to_nmodl(old_statement));
+            for (const auto& replacement: replacement_ptr->second) {
+                logger->debug("SympyReplaceSolutionsVisitor :: adding {}", to_nmodl(replacement));
+            }
+        } else if (to_be_removed_ == nullptr ||
+                   to_be_removed_->find(&(*old_statement)) == to_be_removed_->end()) {
+            new_statements.emplace_back(std::move(old_statement));
+        } else {
+            logger->debug("SympyReplaceSolutionsVisitor :: erasing {}", to_nmodl(old_statement));
+        }
+    }
```

Comment:
why do you do this on every level and not only root?

Reply:
Let's say that I have a statement block inside a statement block (or any other block). If I replace only in the root, while I pass I find in the statementVector the full statement block (which does not match any replacement since the remplacements are done diffEq by diffEquation) and I pass over without replacing anything.